packages/marketo-form/form/index.tsx

```diff
@@ -0,0 +1,137 @@
+    string,
+    {
+      fields: string[]
+      component: (props: { fields: MarketoFormField[] }) => JSX.Element
+    }
+  >
+  components?: MarketoFormikComponents
+  submitTitle?: string
+  className?: string
+  onSubmitSuccess?: () => void
+  onSubmitError?: () => void
+}
+
+const defaultFieldGroupings = {
+  name: {
+    fields: ['FirstName', 'LastName'],
+    component: NameField,
+  },
+}
+
+const Form = ({
+  formId,
+  marketoForm,
+  groups = defaultFieldGroupings,
+  components,
+  submitTitle,
+  className,
+  onSubmitSuccess,
+  onSubmitError,
+}: Props) => {
+  // memoized function that groups marketo fields based on supplied groups.
+  // If a field doesn't belong to a group, it is placed in a group keyed by
+  // field.Name.
+  const groupedFields = useMemo(() => {
+    return groupFields(groups, marketoForm.result)
+  }, [marketoForm, groups])
+
+  const methods = useForm({
+    mode: 'onBlur',
+    defaultValues: calculateDefaultValues(marketoForm.result),
+  })
+
+  const onSubmit = async (data: Record<string, unknown>) => {
+    const res = await fetch(`${window.location.origin}/api/marketo/submit`, {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify({
+        input: [
+          {
+            leadFormFields: convertToRESTFields(data),
+            visitorData: {
+              pageURL: window.location.href,
+            },
+          },
+        ],
+        formId,
+      }),
+    })
+    const marketoResponse = (await res.json()) as { success: boolean }
+    if (res.status === 200 && marketoResponse.success) {
+      if (onSubmitSuccess) {
+        onSubmitSuccess()
+      }
+    } else {
+      console.error(marketoResponse)
+      if (onSubmitError) {
+        onSubmitError()
+      }
+    }
+  }
+
+  return (
+    <FormProvider {...methods}>
+      <form className={className} onSubmit={methods.handleSubmit(onSubmit)}>
+        {Object.entries(groupedFields).map(([groupName, fields]) => {
+          if (groups && groupName in groups) {
+            const Component = groups[groupName].component
+            return (
+              <Component
+                key={fields.map((f) => f.id).join('-')}
+                fields={fields}
+              />
+            )
+          }
+          if (fields.length > 1) {
+            console.warn(
+              `Multiple fields in group ${groupName} with no component defined.`
+            )
+          }
+          return (
+            <Field
+              key={fields[0].id}
+              components={components}
+              field={fields[0]}
+            />
+          )
+        })}
+        <Button
+          disabled={methods.formState.isSubmitting}
+          title={
+            methods.formState.isSubmitting
+              ? 'Submitting...'
+              : submitTitle ?? 'Submit'
+          }
+        />
+        <div>
+          <small>{`Marketo Form (${formId})`}</small>
+        </div>
```

Comment:
I'm guessing this is for testing? 💡 It could be nice to have this only render in dev, if it's something that would be helpful long-term.

Reply:
Yup. I'll move this to a `data` attribute on the form moving forward so that we can still easily confirm the destination Marketo form even in production.